### ISSUE TITLE
Class AST modification proposals

### ISSUE THREAD

**Issue by RReverser** [estree/estree#28]
Collected those from comments to #19, some thoughts on what we could change until it's too late:

### 1. `Class::body::body` -> `Class::body::methods`

```js
interface ClassBody <: Node {
    type: "ClassBody";
    body: [ MethodDefinition ];
```

How about renaming this to `methods` instead? For `BlockStatement`, we can't change it, but here I think we could fix the `classStmt.body[0].body[1].value` chain into smth more meaningful.

### 2. `""` in `MethodDefinition::kind`

```js
interface MethodDefinition <: Node {
    type: "MethodDefinition";
    key: Identifier;
    value: FunctionExpression;
    kind: "" | "get" | "set";
```

How about using `"method"` here instead of empty string?

### 3. `constructor`

Do you think we need to parse constructor as separate node type (i.e. `ConstructorDefinition` or smth) or at least separate `MethodDefinition::kind` or do we continue to treat it as regular method and leave AST processors to check `method.key.type === 'Identifier' && method.key.name === 'constructor'`?

Feels inconvenient as it has different semantics than regular method.

Thoughts?

@dherman @ariya @sebmck @michaelficarra

**Comment by sebmck** [comment]
:+1: for a way to differentiate a class constructor. Semantics are already different inside it `super();` etc so it makes sense to distinguish it from other methods.

**Comment by nzakas** [comment]
Regarding point 1: are we sure that class properties aren't in the future? If they might be coming down the road, changing `body` to `methods` would be prohibitive.

+1 to using "method" for `kind`

**Comment by sebmck** [comment]
Agree with @nzakas. You could add a `properties` property but then you couldn't retain the order which is useful in a lot of scenarios.

**Comment by mikesherov** [comment]
FWIW, only 2. makes sense here to me. `methods` is future hostile, and the check for a constructor is trivial. I'd prefer being able to check all methods including constructor easily over having to know that there is a new mode type specifically for constructors.

**Comment by sebmck** [comment]
@mikesherov It's not really trivial since:

```javascript
class Foo { constructor() {} }
```

and

```javascript
class Foo { "constructor"() {} }
```

are both *the same*. The parser has to check if it's a constructor anyway to implement the `super();` semantics so it'd be convenient to expose it instead of having to do `(method.key.type === "Identifier" && method.key.name === "constructor") || (method.key.type === "Literal" && method.key.value === "constructor")`.

**Comment by mikesherov** [comment]
Yeah, I **suppose** the quoted constructor edge case matters... but I dunno how often it's hit in the real world. 

I suppose if anything, I'd still want it to be the same node type as it is now, and would only really want the `kind` to change, if anything.

**Comment by mikesherov** [comment]
The esprima team discussed this in our meeting today: http://irc.jquery.org/%23esprima-meeting/default_%23esprima-meeting_20150218.log.html#t11:10:19

The conclusion we reached is: 

1. `Class::body::body -> Class::body::methods` is future hostile so we'd like to keep it even if that seems less clear.
2. We're in favor of `"" in MethodDefinition::kind` being changed to `method`.
3. We're in favor of adding a new `MethodDefinition::kind` of `constructor`, and are strongly opposed to a new node type.

@sebmck @RReverser @dherman can we move forward on this? Esprima wants to ship classes soon and we agree that 2 and 3 need addressing.

**Comment by RReverser** [comment]
Sounds good to me, waiting for responses from others and will create PR.

**Comment by mikesherov** [comment]
@rreverser, we've landed these changes in Esprima master. Perhaps open that PR now and get @dherman to :+1:?

**Comment by RReverser** [comment]
Ah, I commited, but didn't open PR. One sec.

**Comment by mikesherov** [comment]
This is resolved now with #33
